### PR TITLE
feat(xcvm): ibc integration spec

### PR DESCRIPTION
**Pull request by hussein-aitlahcen**

### CODE REVIEW

**Review by KaiserKarel — APPROVED** [review]

File: code/xcvm/ibc/INTEGRATION.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# IBC Integration for XCVM
+
+Composable XCVM is a cross-chain virtual machine that is transport and execution
+layer agnostic. One of it's execution layer is CosmWasm and one of it's
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:4]:
```suggestion
layer agnostic. One of it's execution layers is CosmWasm and one of it's
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+transport layer is IBC. CosmWasm is IBC native and provide facilities
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:5]:
```suggestion
transport layers is IBC. CosmWasm is IBC native and provide facilities
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+(functions, metadatas) to leverage IBC in cross-chain communications. We support
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:6]: Plural of data is also data.
```suggestion
(functions, metadata) to leverage IBC in cross-chain communications. We support
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+CosmWasm as smart contract layering on our Picasso parachain.
+
+In the context of the parachain, we decided to split the CosmWasm implementation
+between two components. First, an abstract host agnostic VM that can be
+implemented by any infrastructure (as long as it satisfies the host interface).
+The second component is a Substrate (polkadot framework) Pallet (or module) that
+implement the host functions and instantiate the abstract VM to execute contracts.
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:13]:
```suggestion
implements the host functions and instantiates the abstract VM to execute contracts.
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+## A. CosmWasm VM extension for IBC
+
+Composable is building a CosmWasm VM in an incremental way. To do so, we
+initially chose not to implements the IBC/Staking/Distribution features.
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:18]:
```suggestion
initially choose not to implement the IBC/Staking/Distribution features.
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+submit an IBC packet. Similarly, the IBC callbacks defined by a contract on
+channel opening, packet acknowledgement would never be called (as per the
+previous statement).
+
+To enable this feature, we should extend the
+[`CosmosMsg`](https://github.com/CosmWasm/cosmwasm/blob/531ecc3d942af2040a3a2ce57db9a449110349c7/packages/std/src/results/cosmos_msg.rs#L25)
+data structures to support the
+[`IbcMsg`](https://github.com/CosmWasm/cosmwasm/blob/ded2c78d57b40ac050892b6253ec0a9235246ea5/packages/std/src/ibc.rs#L23)
+variant. This will allow contracts to submit packets and transfer assets via the
+host, assumming it support IBC.
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:40]:
```suggestion
host, assuming it supports IBC.
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+Similarly, we need to extend the VM abstract host to allow it to call back
+contracts on IBC events. The list of events a contract can listen to is [defined
+in the original CosmWasm
+specification](https://github.com/CosmWasm/cosmwasm/blob/531ecc3d942af2040a3a2ce57db9a449110349c7/IBC.md)
+This ability will be implemented by extending the [`VMBase`
+interface](https://github.com/ComposableFi/cosmwasm-vm/blob/24d22367af7602aecc84da390f6f22f88f35b6bb/vm/src/vm.rs#L141) for IBC operations.
+
+## B. Pallet CosmWasm extensions for IBC
+
+Now that the abstract CosmWasm VM is IBC capabable, we still need to upgrade the
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:53]:
```suggestion
Now that the abstract CosmWasm VM is IBC capable, we still need to upgrade the
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+host to implements the extra functions required to keep satifying the interface.
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:54]:
```suggestion
host to implement the extra functions required to keep satifying the interface.
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+And finally, we need to consume the IBC Pallet from the CosmWasm Pallet to
+forward IBC events from the underlying VM to the IBC Pallet and call backs
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:56]:
```suggestion
forward IBC events from the underlying VM to the IBC Pallet and calls back
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+the IBC Pallet as dependency of the CosmWasm Pallet and try interacting with the
+interface to execute IBC operations. If the interface is satisfying and comply
+to the CosmWasm Pallet requirements, we can tie the IBC Pallet interface with
+the CosmWasm VM host interface (the CosmWasm Pallet would act as a router
+between the underlying VM and the IBC pallet).
+
+#### 3. Tying IBC Pallet with the CosmWasm VM
+
+Once the IBC Pallet is a dependency of the CosWasm Pallet and it's interface is
+compatibly with our expectations, we need to tie the IBC Pallet interface with
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:80]:
```suggestion
compatible with our expectations, we need to tie the IBC Pallet interface with
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+the CosmWasm Pallet Host implementation, effectively making the functions
+defined in section **B.1.** not stub anymore. Consequently, we will also be able to
+extend the CosmWasm Pallet benchmarking to cover the weights of this functions.
+
+#### 4. IBC callbacks for contracts
+
+The IBC specification state that the IBC module is only consumed. It does not
+mention how modules consuming IBC can be called back when events occurs (channel
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:88]:
```suggestion
mention how modules consuming IBC can be called back when events occur (channel
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+opened, closed, packet sent...).
+
+It is unclear how we are going to poll events from the IBC Pallet to the
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:91]:
```suggestion
It is unclear how we are going to poll events from the IBC Pallet to
```

File: code/xcvm/ibc/INTEGRATION.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+dispatch them to the according contracts. It can probably done with
```
Comment (aeryz) [inline code/xcvm/ibc/INTEGRATION.md:92]:
```suggestion
dispatch them to the according contracts. It can be probably done with
```

**Review by aeryz — CHANGES_REQUESTED** [review]
I think it clearly defines all the necessary steps to add support for IBC :):)

**Review by aeryz — APPROVED** [review]

**Pull request merged**